ecommerce/extensions/refund/models.py

```
@@ -0,0 +1,65 @@
+from django.conf import settings
+from django.db import models
+from django.utils.translation import ugettext_lazy as _
+from simple_history.models import HistoricalRecords
+
+from ecommerce.extensions.refund.exceptions import InvalidStatus
+
+
+class StatusMixin(object):
+    pipeline_setting = None
+
+    @property
+    def pipeline(self):
+        # NOTE: We use the property and getattr (instead of settings.XXX) so that we can properly override the
+        # settings when testing.
+        return getattr(settings, self.pipeline_setting)
+
+    def available_statuses(self):
+        """ Returns all possible statuses that this object can move to. """
+        return self.pipeline.get(self.status, ())
+
+    # pylint: disable=access-member-before-definition,attribute-defined-outside-init
+    def set_status(self, new_status):
+        """
+        Set a new status for this object.
+
+        If the requested status is not valid, then ``InvalidStatus`` is raised.
+        """
+        if new_status == self.status:
+            return
```

I'm basically with @rlucioni on this - if our model only expects certain state transitions then we should fail loudly when an unexpected transition happens, unless doing so would break oscar somehow.